#### *How data scientists use Python* Task Responses  
1.  - I learned about the differences between software development and data science.
    - I learned about the many different environments and platforms where you can write Python.
    - I learned about the incompatibility of moving from Python 2 to Python 3.
2. The podcast was a great introduction to Python and Data Science. The topic of documenting code is something I very much agree with. There have been times when I didn't document/comment my code and it took me much longer than it should have to refresh myself on the code months later. This is amplified by interacting with code written by others.
3. Yes, it introduced me to some of the many Data Science Python libraries and typical approach to Data Science.
4. The most *surprising* fact for me was how different the approaches to software development and Data Science are. My background is more in Software development, and I am more familiar with class-based programming. I was not aware that Data Science does not typically follow that same appoach.


In [54]:
!wget https://www.gutenberg.org/cache/epub/5827/pg5827.txt -O TheProblemsOfPhilosophy.txt

5971.87s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
--2023-09-27 03:35:33--  https://www.gutenberg.org/cache/epub/5827/pg5827.txt
Resolving www.gutenberg.org (www.gutenberg.org)... 152.19.134.47, 2610:28:3090:3000:0:bad:cafe:47
Connecting to www.gutenberg.org (www.gutenberg.org)|152.19.134.47|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 269684 (263K) [text/plain]
Saving to: ‘TheProblemsOfPhilosophy.txt’

TheProblemsOfPhilos 100%[===================>] 263.36K  --.-KB/s    in 0.1s    

2023-09-27 03:35:33 (1.90 MB/s) - ‘TheProblemsOfPhilosophy.txt’ saved [269684/269684]



In [55]:
from pathlib import Path

file = Path.cwd() / 'TheProblemsOfPhilosophy.txt'

with open(file, 'r') as textFile:
    text = textFile.read()

list = text.split()
list.sort()

In [56]:
from string import punctuation

allWords = []
allUppercaseWords = []

for word in list:
    newWord = word.strip(punctuation).strip("“").strip("”").strip("™").strip("\ufeff").strip("‘").strip("’").strip("•")
    if newWord != '':
        allWords.append(newWord.lower())
        if newWord[0].isupper():
            allUppercaseWords.append(newWord)

allWords.sort()
allUppercaseWords.sort()

In [57]:
import pandas as pd

pd.DataFrame(allWords, columns=['word']).value_counts(sort=False).to_frame('count').to_csv('all_words.csv')
pd.DataFrame(allUppercaseWords, columns=['word']).value_counts(sort=False).to_frame('count').to_csv('all_uppercase_words.csv')

#### *The Problems of Philosophy* Task Responses  
1. The five most frequent words in all_words.csv were:
    - the  
    - of  
    - is 
    - to  
    - and
2. The five most frequent words in all_uppercase_words.csv were:
    - The  
    - I  
    - It  
    - But  
    - Thus
3. Most of the words from both lists are common words with little meaning on their own. The biggest difference are words like 'I' and 'Thus'. 'I' should always be capitalized and 'Thus' is a common way to start a sentence, which makes them less likely to be on the all_word list.
4. The most surprising part of both lists is that the word 'a' is not in the top 5 words. To me it is common both as a start of a sentence and in the middle of sentences. Also, 'But' being in the uppercase words surprised me. I don't commonly start sentences with 'But'.

In [58]:
!wget https://raw.githubusercontent.com/stopwords-iso/stopwords-en/master/stopwords-en.txt -O stopwords.txt

5986.42s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
--2023-09-27 03:35:47--  https://raw.githubusercontent.com/stopwords-iso/stopwords-en/master/stopwords-en.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7677 (7.5K) [text/plain]
Saving to: ‘stopwords.txt’

stopwords.txt       100%[===================>]   7.50K  --.-KB/s    in 0s      

2023-09-27 03:35:47 (110 MB/s) - ‘stopwords.txt’ saved [7677/7677]



In [59]:
from pathlib import Path

swFile = Path.cwd() / 'stopwords.txt'

with open(swFile, 'r') as swText:
    swText = swText.read()

swList = swText.split()
swList.sort()
#print(swList)

In [60]:
nonStopwords = []

for awWord in allWords:
    if awWord.lower() not in swList:
        nonStopwords.append(awWord)

In [61]:
import pandas as pd

nonStopwordsCount = len(nonStopwords)

df = pd.DataFrame(nonStopwords, columns=['word']).value_counts(sort=False).to_frame('count')

df['probability'] = df['count'] / nonStopwordsCount

df.to_csv('all_ns_words.csv')

In [62]:
from string import punctuation

sentence1Prob = 0.00
sentence2Prob = 0.00
newList1 = []
newList2 = []

sentence1 = 'If a belief is true, it can be deduced it is universal.'
sentence2 = 'Criticism of knowledge is counter to scientific results.'

list1 = sentence1.split()
list2 = sentence2.split()

for word0 in list1:
    newWord = word0.strip(punctuation).strip("“").strip("”").strip("™").strip("\ufeff").strip("‘").strip("’").strip("•")
    if newWord != '':
        newList1.append(newWord.lower())

for word0 in list2:
    newWord = word0.strip(punctuation).strip("“").strip("”").strip("™").strip("\ufeff").strip("‘").strip("’").strip("•")
    if newWord != '':
        newList2.append(newWord.lower())

for word1 in newList1:
    if word1 in nonStopwords:
        sentence1Prob += df.loc[word1,'probability'][0]
        
for word2 in newList2:
    if word2 in nonStopwords:
        sentence2Prob += df.loc[word2,'probability'][0]

print(sentence1Prob)
print(sentence2Prob)

0.02120535714285714
0.023958333333333335


#### Non-stop words tasks  
1. There are 2984 non-stop words (non-case sensitive).
2. There are 1462 least probable words that only occur once in the *The Problems of Philosophy*. Some of the words include:  
    - avoided  
    - shortly  
    - sweet  
    - warrants
3. The most common words are those you would most expect to see in a text on philosophy. Some examples include knowledge, belief, and truth.
4. "Criticism of knowledge is counter to scientific results." is more likely with a probability of 0.023958333333333335, while "If a belief is true, it can be deduced it is universal." has a probability of 0.02120535714285714.